dali/kernels/common/copy.h

```diff
@@ -55,6 +55,86 @@ void copy(const TensorView<StorageOut, TOut, NDimIn>& out,
   copy<StorageOut, StorageIn>(out.data, in.data, in.num_elements() * sizeof(TOut), stream);
 }
 
+/**
+ * @brief Copy in to out, merging contiguous samples.
+ *
+ * Contiguous samples are merged to reduce number of copies issued and, at the same time,
+ * allow for non-contiguous data.
+ * The function does not do full coalescing - samples are not sorted to discover order
+ *
+ * This function has very relaxed shape verification - samples that are stored in contiguous
+ * memory are merged and those groups need to have the same total length. Empty samples are skipped
+ * and are not treated as discontinuities.
+ */
+template <typename StorageOut, typename TOut, int NDimOut,
+          typename StorageIn, typename TIn, int NDimIn>
+void copy(const TensorListView<StorageOut, TOut, NDimOut> &out,
+          const TensorListView<StorageIn, TIn, NDimIn> &in,
+          cudaStream_t stream = 0) {
+  static_assert(sizeof(TOut) == sizeof(TIn), "Tensor elements must be of equal size!");
+  static_assert(!std::is_const<TOut>::value, "Cannot copy to a tensor of const elements!");
+  int M = in.num_samples();
+  int N = out.num_samples();
+
+  TOut *out_start = nullptr;
+  const TIn *in_start = nullptr;
+  int64_t out_len = 0, in_len = 0;
+  int i, o;
+  for (i = 0, o = 0; i < M && o < N;) {
+    int64_t in_sample_size  = volume(in.shape.tensor_shape_span(i));
+    int64_t out_sample_size = volume(out.shape.tensor_shape_span(o));
+    if (in_sample_size == 0) {
+      i++;
+      continue;
+    }
+    if (out_sample_size == 0) {
+      o++;
+      continue;
+    }
+    if (in.data[i] != in_start + in_len || out.data[o] != out_start + out_len) {
+      // discontinuity detected
+      if (in_len < out_len) {
+        assert(in.data[i] != in_start + in_len);
+        i++;
+        continue;
+      }
+      if (out_len < in_len) {
+        assert(out.data[o] != out_start + out_len);
+        o++;
+        continue;
+      }
+      assert(in_len == out_len && "Groups of contiguous samples must have equal length");
+      if (out_len)
```

Comment:
```suggestion
      if (out_len > 0)
```
reads better in my opinion